README.md
# Bobbin-CLI

bobbin-cli is a tool designed to make it easy to build, deploy, test and debug embedded
devices using a unified CLI. bobbin-cli understands Rust's cargo / xargo package managers
but can also work with Make or any other build system.

bobbin-cli has the following main areas of functionality:

-  Device enumeration and selection. bobbin-cli recognizes many types of USB debuggers and loaders
   and allows you to set per-project filters so that it knows which device to use even when
   multiple devices are connected to your computer.

-  Build management. bobbin-cli automatically uses xargo to build your project, and reads the
   command line parameters and your Cargo.toml file to automatically determine the output binary
   to use. You can also use Make, with a required parameter to specify the output binary path.

-  Deployment. For supported devices, bobbin-cli can automatically use the appropriate flash
   loading tool (OpenOCD, JLinkExe, bossac or teensy_cli_loader) to upload the output binary.

-  Testing and Debugging. bobbin-cli can automatically connect to and display the virtual serial
   console of the selected device if available. You can also start an instance of OpenOCD and gdb with
   the output binary produced by the build stage.

For a collection of LED blinking demos for a wide variety of popular development boards,
see [bobbin-blinky](https://github.com/bobbin-rs/bobbin-blinky/).

## Host Platforms

MacOS and Linux are currently supported, Windows support is planned.

## Supported Devices

### Debug Probes

Currently Supported:

- [J-Link](https://www.segger.com/products/debug-probes/j-link/) - including CDC (virtual serial port) support.
- [ST-Link/V2](http://www.st.com/en/development-tools/st-link-v2.html) - V2 and V2.1 devices supported,
including CDC (virtual serial port) and SWO Trace (optional, requires libusb).
- [CMSIS-DAP](https://developer.mbed.org/handbook/CMSIS-DAP) - including CDC (virtual serial port) support.
- [DAPLINK](https://github.com/mbedmicro/DAPLink) - including MSD (mass storage device) and CDC (virtual serial port) support.
- [TI ICDI](http://www.ti.com/tool/stellaris_icdi_drivers) - including CDC (virtual serial port) support.
- [Black Magic Probe](https://github.com/blacksphere/blackmagic) - including CDC (virtual serial port) support.

Coming Soon:

- [PEmicro](http://www.pemicro.com/opensda/)

### Development Boards with Embedded Debug Probes

Boards from the following product families include embedded debug probes that should be supported.

- [STM32 Discovery](http://www.st.com/en/evaluation-tools/stm32-mcu-discovery-kits.html?querycriteria=productId=LN1848) 
- [STM32 Nucleo](http://www.st.com/en/evaluation-tools/stm32-mcu-nucleo.html?querycriteria=productId=LN1847)
- [Freescale FRDM](http://www.nxp.com/products/software-and-tools/hardware-development-tools/freedom-development-boards:FREDEVPLA)
- [TI Launchpad](http://www.ti.com/lsds/ti/tools-software/launchpads/overview/overview.page)
- [NXP S32K144EVB](http://www.nxp.com/products/automotive-products/microcontrollers-and-processors/arm-mcus-and-mpus/s32-processors-microcontrollers/s32k144-evaluation-board:S32K144EVB)
- [Arduino Zero](https://www.arduino.cc/en/Guide/ArduinoZero)

*Note:* Many development boards support OpenSDA, which allows a choice of firmware to be installed. Debug probes may support
CMSIS-DAP, DAPLINK, J-Link and PEMicro firmware variants. Be sure to upgrade to the most recent firmware available, and ensure that
a variant supporting OpenOCD (CMSIS-DAP/DAPLINK) or J-Link is installed. Please see [Upgrading Development Board Firmware](FIRMWARE.md).

### Development Boards with Flash Loader Support

Boards from the following product families use flash loaders that are supported.

- [Feather M0](https://www.adafruit.com/feather)
- [Teensy 3.x and LC](https://www.pjrc.com/teensy/)
- STM32 DFU Bootloader

## Prerequisites

### Build Tools

These tools must be installed in your PATH.

- [GNU ARM Embedded](https://developer.arm.com/open-source/gnu-toolchain/gnu-rm)
- [xargo](https://github.com/japaric/xargo)

### Debugger / Loader Tools

You must have the appropriate tools installed for the debug probes / dev boards that you wish to use.

- [OpenOCD](http://openocd.org) - 0.10 or later required for STLink, DAPLINK, CMSIS-DAP, and TI ICDI debug probes
- [J-Link](https://www.segger.com/downloads/jlink) - required for J-Link debug probes.
- [Bossa](http://www.shumatech.com/web/products/bossa) - required for Arduino and Feather devices
- [Teensy Loader](https://www.pjrc.com/teensy/loader_cli.html) - required for Teensy devices
- [libusb](http://libusb.info) - required for STLink SWO Trace support.
- [dfu-util](http://dfu-util.sourceforge.net) - required for STM32 DFU Bootloader support

### Development Board Firmware

If you are using a development board with embedded debug probe, check that you know what debug firmware you have
installed and that it is up to date. Please see [Development Board Firmware](FIRMWARE.md).

## Installation

*Note:* Only Linux and macOS hosts are supported at this time.

To install from cargo:

```
$ cargo install bobbin-cli
```

To install from github:

```
$ git clone https://github.com/bobbin-rs/bobbin-cli.git
$ cd bobbin-cli
$ cargo install
```

To install with ST-Link SWV Trace support:

```
$ cargo install --features stlink
```

## Usage

The name of the executable is `bobbin`.

### Help

You can display detailed help text by using `bobbin -h`.

### Bobbin Check

Use "bobbin check" to list the version numbers of all Bobbin dependencies. "Not Found" will be displayed if
the dependency is not available.

```
$ bobbin check
      Rust 1.20.0-nightly (83c659ef6 2017-07-18)
     Cargo 0.21.0-nightly (f709c35a3 2017-07-13)
     Xargo 0.3.8
       GCC 5.4.1 20160919 (release) [ARM/embedded-5-branch revision 240496]
   OpenOCD 0.10.0+dev-00092-g77189db (2017-03-01-20:42)
     JLink V6.15c (Compiled Apr 24 2017 19:07:08)
     Bossa 1.7.0
    Teensy 2.1
  dfu-util 0.9
```

Please include the "bobbin check" output when reporting problems.

### Bobbin List

Use "bobbin list" to view all debug probes and development boards connected to your host.

```
$ bobbin list
ID       VID :PID  Vendor                   Product                          Serial Number
4c01a4ad 1366:0105 SEGGER                   J-Link                           000621000000
14a7f5da 03eb:2157 Atmel Corp.              EDBG CMSIS-DAP                   00000000EZE000005574
b7e67550 0483:374b STMicroelectronics       STM32 STLink                     0673FF485550755187121723
a3ef65e3 0483:374b STMicroelectronics       STM32 STLink                     0667FF555654725187073723
cb46720d 1cbe:00fd Texas Instruments        In-Circuit Debug Interface       0F007E1A
8c6bbec5 0d28:0204 ARM                      DAPLink CMSIS-DAP                0260000025414e450049501247e0004e30f1000097969900
f95f4aca 0d28:0204 ARM                      DAPLink CMSIS-DAP                0240000034544e45001b00028aa9001a2011000097969900
c2f3dc42 0483:374b STMicroelectronics       STM32 STLink                     0670FF484957847167071621
$
```

The device ID is a hash of the USB Vendor ID, USB Product ID, and USB Serial Number (if available). "bobbin list" displays
the first eight hex digits of the device ID, and "bobbin info" displays the full 64 bit ID.

### Bobbin Info

To view detailed information about a devices, use the "bobbin info" subcommand.

```
$ bobbin -d 4c01 info
ID               c2f3dc42b4aadc58b6dfa98ce527dd436e3e4fa5
Vendor ID        0483
Product ID       374b
Vendor           STMicroelectronics
Product          STM32 STLink
Serial Number    0670FF484957847167071621
Type             STLinkV21
Loader Type      OpenOCD
Debugger Type    OpenOCD
CDC Device       /dev/cu.usbmodem141413
OpenOCD Serial   hla_serial 0670FF484957847167071621
$
```

If you have more than one connected device, you can select a specific device by using the -d command line
parameter. bobbin-cli will also look for a device filter directive in a YAML configuration file at ./bobbin/config

```
$ cat .bobbin/config
[filter]
device = "c2f3dc42"
```

### Bobbin Build

`bobbin build` runs xargo (by default) or make to build your application. If using xargo, bobbin-cli will
pass through any --target, --bin, --example or --release parameters. 

On completion, bobbin-cli will run `arm-none-eabi-size` on the binary and display the output.

```
$ bobbin build
   Compiling blue-pill v0.1.0 (file:///home/bobbin/bobbin-blinky/blue-pill)
    Finished dev [optimized + debuginfo] target(s) in 0.50 secs
   text	   data	    bss	    dec	    hex	filename
    152	      0	      4	    156	     9c	target/thumbv7em-none-eabihf/debug/blue-pill
$
```

### Bobbin Load

`bobbin load` runs `bobbin build` and then, if successful, load the binary onto the device
using the selected debugger or loader, using objcopy as needed to convert to the appropriate
format. You may include --target, --bin, --example or --release parameters which will be passed
to `bobbin build`.

`bobbin load` will interpret the build parameters as well as the Cargo.toml file to determine
the path to the binary.

```
$ bobbin load
   Compiling blue-pill v0.1.0 (file:///home/bobbin/bobbin-blinky/blue-pill)
    Finished dev [optimized + debuginfo] target(s) in 0.13 secs
   text	   data	    bss	    dec	    hex	filename
    152	      0	      4	    156	     9c	target/thumbv7em-none-eabihf/debug/blue-pill
     Loading target/thumbv7em-none-eabihf/debug/blue-pill.hex
    Complete Successfully flashed device
      Loader Load Complete
$
```


Some devices require manual intervention to enter bootloader mode; you should do this before
running `bobbin load`.

*Note: Many debuggers and loaders require additional configuration*

- OpenOCD: a properly configured openocd.cfg file must be in the current directory. bobbin will add 
the appropriate OpenOCD command line parameters to select the specific device.
- J-Link: you must specify the device type through the --jlink-device command line parameter or
in the .bobbin/config file.
- Teensy Loader: you must specify the device type through the --teensy-mcu command line parameter or
in the .bobbin/config file.

### Bobbin Run

`bobbin run` runs `bobbin load` and then, if successful, open the serial console of the
connected device to display the output. Use Control-C to terminate this console viewer.

If the selected device does not have an associated serial port, that step will be skipped. 

You can use the --console parameter to manually specify a serial device, or --noconsole if 
you do not want run the console viewer at all.

*Note: the serial viewer is currently hard-coded to 115,200 baud*

If bobbin-cli is compiled with support for SWO trace, you can pass the --itm parameter
to display ITM output instead of running the serial console. You will also need to pass
the --itm-target-clock parameter with the target's clock speed.

```
$ bobbin run
   Compiling blue-pill v0.1.0 (file:///home/bobbin/bobbin-hello/blue-pill)
    Finished dev [optimized + debuginfo] target(s) in 0.13 secs
   text	   data	    bss	    dec	    hex	filename
    152	      0	      4	    156	     9c	target/thumbv7em-none-eabihf/debug/blue-pill
     Loading target/thumbv7em-none-eabihf/debug/blue-pill.hex
    Complete Successfully flashed device
      Loader Load Complete
     Console Opening Console
Hello World 1
Hello World 2
Hello World 3
^C
$
```

### Bobbin Test

`bobbin test` runs `bobbin run` and then interprets the serial output, looking for
tags indicating test progress and completion.

```
$ bobbin test
   Compiling frdm-k64f v0.1.0 (file:///home/bobbin/bobbin-boards/frdm-k64f)
    Finished dev [optimized + debuginfo] target(s) in 0.61 secs
   text	   data	    bss	    dec	    hex	filename
   6252	    428	    408	   7088	   1bb0	target/thumbv7em-none-eabihf/debug/frdm-k64f
     Loading target/thumbv7em-none-eabihf/debug/frdm-k64f
    Complete Successfully flashed device
      Loader Load Complete
     Console Opening Console
[start] Running tests for frdm-k64f
[pass] 0
[pass] 1
[pass] 2
[pass] 3
[pass] 4
[done] All tests passed
$
```


CHANGELOG
0.9.0:
   Added console options to configuation file
   Infer --target by reading .cargo/configuration   
   Rewrite to use Serde to read configuration files
   Add support for remote operations using SSH
   Also search for openocd.cfg in .bobbin/<device-id>/ and ~/.bobbin/<device-id>/
   Use -mmcu instead of --mcu option for teensy_loader_cli compatibility

src/cmd.rs
use Result;
use clap::ArgMatches;
use config::Config;
use printer::Printer;
use std::io::{self, Read, Write};
use std::path::PathBuf;
use std::process::*;
// use std::os::unix::io::*;
use std::os::unix::process::CommandExt;
// use std::fs::File;

use device;
use builder;
use loader;
use debugger;
use console;
use check;
use tempfile;

pub fn check(
    cfg: &Config,
    args: &ArgMatches,
    cmd_args: &ArgMatches,
    out: &mut Printer,
) -> Result<()> {
    writeln!(out, "    Bobbin {}", crate_version!())?;
    writeln!(out, "      Rust {}", check::rust_version().unwrap_or(String::from("Not Found")))?;
    writeln!(out, "     Cargo {}", check::cargo_version().unwrap_or(String::from("Not Found")))?;
    writeln!(out, "     Xargo {}", check::xargo_version().unwrap_or(String::from("Not Found")))?;
    writeln!(out, "       GCC {}", check::gcc_version().unwrap_or(String::from("Not Found")))?;
    writeln!(out, "   OpenOCD {}", check::openocd_version().unwrap_or(String::from("Not Found")))?;
    writeln!(out, "     JLink {}", check::jlink_version().unwrap_or(String::from("Not Found")))?;
    writeln!(out, "     Bossa {}", check::bossac_version().unwrap_or(String::from("Not Found")))?;    
    writeln!(out, "    Teensy {}", check::teensy_version().unwrap_or(String::from("Not Found")))?;
    writeln!(out, "  dfu-util {}", check::dfu_util_version().unwrap_or(String::from("Not Found")))?;
    Ok(())
}

pub fn list(
    cfg: &Config,
    args: &ArgMatches,
    cmd_args: &ArgMatches,
    out: &mut Printer,
) -> Result<()> {   
    if let Some(host) = args.value_of("host").or_else(|| cfg.filter_host()) {
        let mut cmd = Command::new("ssh");
        cmd.arg("-q");
        cmd.arg("-t");
        cmd.arg(host);
        cmd.arg(".cargo/bin/bobbin");
        if let Some(device) = cfg.device(args) {
            cmd.arg("--device").arg(device);
        }        
        if args.is_present("verbose") {
            cmd.arg("--verbose");
        }                
        cmd.arg("list");
        cmd.exec();
        unreachable!()
    }

    let filter = device::filter(cfg, args, cmd_args);
    let devices = device::search(&filter);

    writeln!(out, "{:08} {:08}  {:40} {:24}",
        "ID",
        " VID:PID",
        "Vendor / Product",
        "Serial Number",
        )?;
    for d in devices?.iter() {
        let u = d.usb();
        write!(out, "{:08} {:04x}:{:04x} {:40} {:24}",
            &d.hash()[..8],
            u.vendor_id,
            u.product_id,
            format!("{} / {}", u.vendor_string, u.product_string),
            u.serial_number,
            )?;
        writeln!(out, "")?;
    }
    Ok(())
}


pub fn info(
    cfg: &Config,
    args: &ArgMatches,
    cmd_args: &ArgMatches,
    out: &mut Printer,
) -> Result<()> {
    if let Some(host) = args.value_of("host").or_else(|| cfg.filter_host()) {
        let mut cmd = Command::new("ssh");
        cmd.arg("-q");
        cmd.arg("-t");
        cmd.arg(host);
        cmd.arg(".cargo/bin/bobbin");
        if let Some(device) = cfg.device(args) {
            cmd.arg("--device").arg(device);
        }        
        if args.is_present("verbose") {
            cmd.arg("--verbose");
        }                
        cmd.arg("info");
        cmd.exec();
        unreachable!()
    }

    let filter = device::filter(cfg, args, cmd_args);
    let devices = device::search(&filter)?;

    for d in devices.iter() {
        let u = d.usb();
        writeln!(out, "{:16} {}", "ID", d.hash())?;
        writeln!(out, "{:16} {:04x}", "Vendor ID", u.vendor_id)?;
        writeln!(out, "{:16} {:04x}", "Product ID", u.product_id)?;
        writeln!(out, "{:16} {}", "Vendor", u.vendor_string)?;
        writeln!(out, "{:16} {}", "Product", u.product_string)?;
        writeln!(out, "{:16} {}", "Serial Number", u.serial_number)?;
        writeln!(
            out,
            "{:16} {}",
            "Type",
            d.device_type().unwrap_or("Unknown")
        )?;
        if let Some(loader_type) = d.loader_type() {
            writeln!(out, "{:16} {}", "Loader Type", loader_type)?;
        }
        if let Some(debugger_type) = d.debugger_type() {
            writeln!(out, "{:16} {}", "Debugger Type", debugger_type)?;
        }

        if let Some(bossa_path) = d.bossa_path() {
            writeln!(out, "{:16} {}", "Bossac Device", bossa_path)?;
        }
        if let Some(cdc_path) = d.cdc_path() {
            writeln!(out, "{:16} {}", "CDC Device", cdc_path)?;
        }
        if let Some(msd_path) = d.msd_path() {
            writeln!(out, "{:16} {}", "MSD Device", msd_path.display())?;
        }
        if let Some(gdb_path) = d.gdb_path() {
            writeln!(out, "{:16} {}", "GDB Device", gdb_path)?;
        }        
        if let Some(openocd_serial) = d.openocd_serial() {
            writeln!(out, "{:16} {}", "OpenOCD Serial", openocd_serial)?;
        }
        writeln!(out, "")?;
    }
    Ok(())
}

pub fn build(
    cfg: &Config,
    args: &ArgMatches,
    cmd_args: &ArgMatches,
    out: &mut Printer,
) -> Result<()> {
    let dst = builder::build(cfg, args, args, out)?;
    Ok(())
}

pub fn load(
    cfg: &Config,
    args: &ArgMatches,
    cmd_args: &ArgMatches,
    out: &mut Printer,
) -> Result<()> {

    let dst = if let Some(dst) = builder::build(cfg, args, cmd_args, out)? {
        dst
    } else {
        bail!("No build output available to load");
    };

    if let Some(host) = args.value_of("host").or_else(|| cfg.filter_host()) {
        let mut cmd = Command::new("rsync");
        cmd.arg(dst.clone());
        let device = args.value_of("device").or_else(|| cfg.filter_device()).unwrap_or_else(|| "bobbin");
        cmd.arg(format!("{}:/tmp/{}/", host, device));
        let status = cmd.status()?;
        if !status.success() {
            bail!("Unable to transfer file to {}", host);
        }
        let mut cmd = Command::new("ssh");
        cmd.arg("-q");
        cmd.arg("-t");
        cmd.arg(host);
        cmd.arg(".cargo/bin/bobbin");
        if let Some(device) = cfg.device(args) {
            cmd.arg("--device").arg(device);
        }        
        if args.is_present("verbose") {
            cmd.arg("--verbose");
        }          
        let subcmd = if args.is_present("load") {
            "load"
        } else if args.is_present("run") {
            "run"
        } else if args.is_present("test") {
            "test"
        } else {
            bail!("Only load, run and test are supported for remote hosts")
        };
        cmd.arg(subcmd);

        if let Some(arg) = cfg.jlink_device(cmd_args) {
            cmd.arg("--jlink-device").arg(arg);
        }

        if let Some(arg) = cfg.teensy_mcu(cmd_args) {
            cmd.arg("--teensy-mcu").arg(arg);
        }

        if let Some(arg) = cfg.blackmagic_mode(cmd_args) {
            cmd.arg("--blackmagic_mode").arg(arg);
        }

        if let Some(arg) = cfg.console(cmd_args) {
            cmd.arg("--console").arg(arg);
        }

        cmd.arg(format!("/tmp/{}/{}", device, dst.file_name().unwrap().to_str().unwrap()));
        out.verbose("Remote", &format!("{:?}", cmd))?;

        cmd.exec();
        unreachable!()
    }

    let filter = device::filter(cfg, args, cmd_args);
    let mut devices = device::search(&filter)?;

    let device = if devices.len() == 0 {
        bail!("No matching devices found.");
    } else if devices.len() > 1 {
        bail!("More than one device found ({})", devices.len());
    } else {
        devices.remove(0)
    };

    let ldr = if let Some(ldr) = device.loader_type() {
        out.verbose("loader", ldr)?;
        if let Some(ldr) = loader::loader(ldr) {
            ldr
        } else {
            bail!("Unknown loader type: {}", ldr);
        }
    } else {
        bail!("Selected device has no associated loader");
    };

    let con = if !cmd_args.is_present("noconsole") && !cmd_args.is_present("itm") {
        if args.is_present("run") || args.is_present("test") {
            if let Some(cdc_path) = cfg.console(cmd_args) {
                let mut con = console::open(&cdc_path)?;
                con.clear()?;
                Some(con)
            } else if let Some(cdc_path) = device.cdc_path() {
                let mut con = console::open(&cdc_path)?;
                con.clear()?;
                Some(con)
            } else {
                None
            }
        } else {
            None
        }
    } else {
        None
    };

    if dst == PathBuf::from("--") {
        let mut buffer: Vec<u8> = Vec::new();
        io::stdin().read_to_end(&mut buffer)?;
        out.verbose("stdin", &format!("Read {} bytes from stdin", buffer.len()))?;
        let mut tmpfile = tempfile::NamedTempFile::new()?;
        tmpfile.write(buffer.as_ref())?;
        tmpfile.flush()?;
        ldr.load(
            cfg,
            args,
            cmd_args,
            out,
            device.as_ref(),
            tmpfile.path()
        )?;
        out.verbose("stdin", "Removing temporary file.")?;
    } else {
        ldr.load(
            cfg,
            args,
            cmd_args,
            out,
            device.as_ref(),
            dst.as_path(),
        )?;
    }
    out.info("Loader", "Load Complete")?;

    if cmd_args.is_present("itm") {
        if device.can_trace_itm() {
            out.info("ITM", "Starting ITM Trace")?;
            let target_clk = if let Some(v) = cmd_args.value_of("itm-target-clock") {
                v.parse::<u32>()?
            } else {
                if let Some(v) = cfg.itm_target_clock() {
                    v
                } else {
                    bail!("itm-target-clock is required for ITM trace.")
                }
            };
            let trace_clk = 2_000_000;
            device.trace_itm(target_clk, trace_clk)?;

        } else {
            bail!("Currently selected device does not support ITM trace");
        }
    } else if let Some(mut con) = con {
        out.info("Console", "Opening Console")?;
        if args.is_present("test") {
            con.test(&args, &cmd_args)?;
        } else {
            con.view()?;
        }
    }

    Ok(())
}

pub fn control(
    cfg: &Config,
    args: &ArgMatches,
    cmd_args: &ArgMatches,
    out: &mut Printer,
) -> Result<()> {
    if let Some(host) = args.value_of("host").or_else(|| cfg.filter_host()) {
        let mut cmd = Command::new("ssh");
        cmd.arg("-q");
        cmd.arg("-t");
        cmd.arg(host);
        cmd.arg(".cargo/bin/bobbin");
        if let Some(device) = cfg.device(args) {
            cmd.arg("--device").arg(device);
        }        
        if args.is_present("verbose") {
            cmd.arg("--verbose");
        }          
        let subcmd = if args.is_present("halt") {
            cmd.arg("halt");
        } else if args.is_present("resume") {
            cmd.arg("resume");
        } else if args.is_present("reset") {
            cmd.arg("reset");
            if cmd_args.is_present("run") {
                cmd.arg("--run");
            } else if cmd_args.is_present("halt") {
                cmd.arg("--halt");
            } else if cmd_args.is_present("init") {
                cmd.arg("--init");
            }        
        };

        if let Some(arg) = cfg.jlink_device(cmd_args) {
            cmd.arg("--jlink-device").arg(arg);
        }

        if let Some(arg) = cfg.teensy_mcu(cmd_args) {
            cmd.arg("--teensy-mcu").arg(arg);
        }

        if let Some(arg) = cfg.blackmagic_mode(cmd_args) {
            cmd.arg("--blackmagic_mode").arg(arg);
        }

        if let Some(arg) = cfg.console(cmd_args) {
            cmd.arg("--console").arg(arg);
        }        
        out.verbose("Remote", &format!("{:?}", cmd))?;

        cmd.exec();
        unreachable!()
    }

    let filter = device::filter(cfg, args, cmd_args);
    let mut devices = device::search(&filter)?;

    let device = if devices.len() == 0 {
        bail!("No matching devices found.");
    } else if devices.len() > 1 {
        bail!("More than one device found ({})", devices.len());
    } else {
        devices.remove(0)
    };

    let dbg = if let Some(dbg) = device.debugger_type() {
        out.verbose("debugger", dbg)?;
        if let Some(dbg) = debugger::debugger(dbg) {
            dbg
        } else {
            bail!("Unknown debugger type: {}", dbg);
        }
    } else {
        bail!("Selected device has no associated loader");
    };

    if let Some(_) = args.subcommand_matches("halt") {
        dbg.halt(cfg, args, cmd_args, out, device.as_ref())?;
    } else if let Some(_) = args.subcommand_matches("resume") {
        dbg.resume(cfg, args, cmd_args, out, device.as_ref())?;
    } else if let Some(_) = args.subcommand_matches("reset") {
        if cmd_args.is_present("run") {
            dbg.reset_run(cfg, args, cmd_args, out, device.as_ref())?;
        } else if cmd_args.is_present("halt") {
            dbg.reset_halt(cfg, args, cmd_args, out, device.as_ref())?;
        } else if cmd_args.is_present("init") {
            dbg.reset_init(cfg, args, cmd_args, out, device.as_ref())?;
        } else {
            dbg.reset(cfg, args, cmd_args, out, device.as_ref())?;
        }
    }

    Ok(())
}

pub fn openocd(
    cfg: &Config,
    args: &ArgMatches,
    cmd_args: &ArgMatches,
    out: &mut Printer,
) -> Result<()> {
    if let Some(host) = args.value_of("host").or_else(|| cfg.filter_host()) {
        let mut cmd = Command::new("ssh");
        cmd.arg("-q");
        cmd.arg("-t");
        cmd.arg("-L").arg("3333:localhost:3333");        
        cmd.arg(host);
        cmd.arg(".cargo/bin/bobbin");
        if let Some(device) = cfg.device(args) {
            cmd.arg("--device").arg(device);
        }        
        if args.is_present("verbose") {
            cmd.arg("--verbose");
        }                
        cmd.arg("openocd");
        println!("{:?}", cmd);
        cmd.exec();
        unreachable!()
    }

    let filter = device::filter(cfg, args, cmd_args);
    let mut devices = device::search(&filter)?;

    let device = if devices.len() == 0 {
        bail!("No matching devices found.");
    } else if devices.len() > 1 {
        bail!("More than one device found ({})", devices.len());
    } else {
        devices.remove(0)
    };

    let dbg = debugger::OpenOcdDebugger {};
    dbg.run(cfg, args, cmd_args, out, device.as_ref())?;
    unreachable!()
}


pub fn jlink(
    cfg: &Config,
    args: &ArgMatches,
    cmd_args: &ArgMatches,
    out: &mut Printer,
) -> Result<()> {
    if let Some(host) = args.value_of("host").or_else(|| cfg.filter_host()) {
        let mut cmd = Command::new("ssh");
        cmd.arg("-q");
        cmd.arg("-t");
        cmd.arg("-L").arg("3333:localhost:3333");
        cmd.arg(host);
        cmd.arg(".cargo/bin/bobbin");
        if let Some(device) = cfg.device(args) {
            cmd.arg("--device").arg(device);
        }        
        if args.is_present("verbose") {
            cmd.arg("--verbose");
        }                
        cmd.arg("jlink");
        if let Some(arg) = cfg.jlink_device(cmd_args) {
            cmd.arg("--jlink-device").arg(arg);
        }        
        cmd.exec();
        unreachable!()
    }

    let filter = device::filter(cfg, args, cmd_args);
    let mut devices = device::search(&filter)?;

    let device = if devices.len() == 0 {
        bail!("No matching devices found.");
    } else if devices.len() > 1 {
        bail!("More than one device found ({})", devices.len());
    } else {
        devices.remove(0)
    };

    let jlink_dev = if let Some(jlink_dev) = cfg.jlink_device(cmd_args) {
        jlink_dev
    } else {
        bail!("JLink Loader requires that --jlink-device is specified");
    }; 

    let mut cmd = Command::new("JLinkGDBServer");
    cmd.arg("-device").arg(jlink_dev);
    cmd.arg("-if").arg("SWD");
    cmd.arg("-speed").arg("4000");
    cmd.arg("-port").arg("3333");
    cmd.arg("-select").arg(
        format!("usb={}",device.usb().serial_number),
    );

    cmd.exec();

    let status = cmd.status()?;
    if !status.success() {
        bail!("jlink failed")
    }
    Ok(())
}


pub fn gdb(
    cfg: &Config,
    args: &ArgMatches,
    cmd_args: &ArgMatches,
    out: &mut Printer,
) -> Result<()> {
    let dst = if let Some(dst) = builder::build(cfg, args, cmd_args, out)? {
        dst
    } else {
        bail!("No build output available for gdb");
    };

    if let Some(host) = args.value_of("host").or_else(|| cfg.filter_host()) {
        let mut cmd = Command::new("arm-none-eabi-gdb");
        cmd.arg("-ex").arg(format!("target extended-remote :3333"));
        cmd.arg(dst);
        cmd.exec();
        unreachable!()
    }

    let filter = device::filter(cfg, args, cmd_args);
    let mut devices = device::search(&filter)?;

    let device = if devices.len() == 0 {
        bail!("No matching devices found.");
    } else if devices.len() > 1 {
        bail!("More than one device found ({})", devices.len());
    } else {
        devices.remove(0)
    };

    let mut cmd = Command::new("arm-none-eabi-gdb");
    if let Some(gdb_path) = device.gdb_path() {
        cmd.arg("-ex").arg(format!("target extended-remote {}", gdb_path));
        // These commands are BlackMagic Probe Specific
        cmd.arg("-ex").arg("monitor swdp_scan");
        cmd.arg("-ex").arg("attach 1");
    }
    cmd.stdout(Stdio::inherit()).stderr(Stdio::inherit()).arg(
        dst,
    );
    out.verbose("gdb", &format!("{:?}", cmd))?;

    cmd.exec();

    let status = cmd.status()?;
    if !status.success() {
        bail!("gdb failed")
    }
    Ok(())
}

pub fn console(
    cfg: &Config,
    args: &ArgMatches,
    cmd_args: &ArgMatches,
    out: &mut Printer,
) -> Result<()> {
    if let Some(host) = args.value_of("host").or_else(|| cfg.filter_host()) {
        let mut cmd = Command::new("ssh");
        cmd.arg("-q");
        cmd.arg("-t");
        cmd.arg(host);
        cmd.arg(".cargo/bin/bobbin");
        if let Some(device) = cfg.device(args) {
            cmd.arg("--device").arg(device);
        }        
        if args.is_present("verbose") {
            cmd.arg("--verbose");
        }                
        cmd.arg("console");
        if let Some(arg) = cfg.console(cmd_args) {
            cmd.arg("--console").arg(arg);
        }                
        cmd.exec();
        unreachable!()
    }

    let filter = device::filter(cfg, args, cmd_args);
    let mut devices = device::search(&filter)?;

    let device = if devices.len() == 0 {
        bail!("No matching devices found.");
    } else if devices.len() > 1 {
        bail!("More than one device found ({})", devices.len());
    } else {
        devices.remove(0)
    };

    if let Some(cdc_path) = cfg.console(cmd_args) {
        let mut con = console::open(&cdc_path)?;
        con.view()?
    } else if let Some(cdc_path) = device.cdc_path() {
        let mut con = console::open(&cdc_path)?;
        con.view()?
    } else {
        bail!("No console found for device");
    }

    Ok(())
}

pub fn screen(
    cfg: &Config,
    args: &ArgMatches,
    cmd_args: &ArgMatches,
    out: &mut Printer,
) -> Result<()> {
    if let Some(host) = args.value_of("host").or_else(|| cfg.filter_host()) {
        let mut cmd = Command::new("ssh");
        cmd.arg("-q");
        cmd.arg("-t");
        cmd.arg(host);
        cmd.arg(".cargo/bin/bobbin");
        if let Some(device) = cfg.device(args) {
            cmd.arg("--device").arg(device);
        }        
        if args.is_present("verbose") {
            cmd.arg("--verbose");
        }                
        cmd.arg("screen");
        if let Some(arg) = cfg.console(cmd_args) {
            cmd.arg("--console").arg(arg);
        }                
        cmd.exec();
        unreachable!()
    }

    let filter = device::filter(cfg, args, cmd_args);
    let mut devices = device::search(&filter)?;

    let device = if devices.len() == 0 {
        bail!("No matching devices found.");
    } else if devices.len() > 1 {
        bail!("More than one device found ({})", devices.len());
    } else {
        devices.remove(0)
    };

    let mut cmd = Command::new("screen");
    if let Some(cdc_path) = cfg.console(cmd_args) {
        cmd.arg(cdc_path);
    } else if let Some(cdc_path) = device.cdc_path() {
        cmd.arg(cdc_path);
    } else {
        bail!("No serial device path found");
    }
    cmd.arg("115200");
    cmd.exec();

    let status = cmd.status()?;
    if !status.success() {
        bail!("screen failed")
    }
    Ok(())
}

pub fn objdump(
    cfg: &Config,
    args: &ArgMatches,
    cmd_args: &ArgMatches,
    out: &mut Printer,
) -> Result<()> {
    Ok(())
}

pub fn itm(
    cfg: &Config,
    args: &ArgMatches,
    cmd_args: &ArgMatches,
    out: &mut Printer,
) -> Result<()> {
    if let Some(host) = args.value_of("host").or_else(|| cfg.filter_host()) {
        let mut cmd = Command::new("ssh");
        cmd.arg("-q");
        cmd.arg("-t");
        cmd.arg(host);
        cmd.arg(".cargo/bin/bobbin");
        if let Some(device) = cfg.device(args) {
            cmd.arg("--device").arg(device);
        }        
        if args.is_present("verbose") {
            cmd.arg("--verbose");
        }                
        cmd.arg("itm");
        cmd.exec();
        unreachable!()
    }

    let filter = device::filter(cfg, args, cmd_args);
    let mut devices = device::search(&filter)?;

    let device = if devices.len() == 0 {
        bail!("No matching devices found.");
    } else if devices.len() > 1 {
        bail!("More than one device found ({})", devices.len());
    } else {
        devices.remove(0)
    };

    if device.can_trace_itm() {
        out.info("ITM", "Starting ITM Trace")?;
        let target_clk = if let Some(v) = cmd_args.value_of("itm-target-clock") {
            v.parse::<u32>()?
        } else {
            if let Some(v) = cfg.itm_target_clock() {
                v
            } else {
                bail!("itm-target-clock is required for ITM trace.")
            }
        };
        let trace_clk = 2_000_000;
        device.trace_itm(target_clk, trace_clk)?;
    } else {
        bail!("Currently selected device does not support ITM trace");
    }
    Ok(())
}
